substrate/frame/examples/mock-builder/src/mock/mock_pallets.rs

```diff
@@ -0,0 +1,384 @@
+///
+/// NOTE: It's expected that if you own the trait, you can simply put one procedural attribute on
+/// top of it to autogerate the mocked pallet.
+/// ```ignore
+/// #[mock_builder::pallet]
+/// mod mock_pallet_time {
+///     trait Currency<AccountId> {
+///         type Balance: Balance + MaybeSerializeDeserialize;
+///         type PositiveImbalance: Imbalance<Self::Balance, Opposite = Self::NegativeImbalance>;
+///         type NegativeImbalance: Imbalance<Self::Balance, Opposite = Self::PositiveImbalance>;
+///         fn total_balance(a: &AccountId) -> Self::Balance;
+///         fn can_slash(a: &T::AccountId, b: Self::Balance);
+///         fn total_issuance() -> Self::Balance;
+///         fn minimum_balance() -> Self::Balance;
+///         fn burn(a: Self::Balance) -> Self::PositiveImbalance;
+///         fn issue(a: Self::Balance) -> Self::NegativeImbalance;
+///         fn free_balance(a: &T::AccountId) -> Self::Balance;
+///         fn ensure_can_withdraw(
+///             a: &T::AccountId,
+///             b: Self::Balance,
+///             c: WithdrawReasons,
+///             d: Self::Balance,
+///         ) -> DispatchResult;
+///         fn transfer(
+///             a: &T::AccountId,
+///             b: &T::AccountId,
+///             c: Self::Balance,
+///             d: ExistenceRequirement,
+///         ) -> DispatchResult;
+///         fn slash(a: &T::AccountId, b: Self::Balance) -> (Self::NegativeImbalance, Self::Balance);
+///         fn deposit_into_existing(
+///             a: &T::AccountId,
+///             b: Self::Balance,
+///         ) -> Result<Self::PositiveImbalance, DispatchError>;
+///         fn deposit_creating(a: &T::AccountId, b: Self::Balance) -> Self::PositiveImbalance;
+///         fn withdraw(
+///             a: &T::AccountId,
+///             b: Self::Balance,
+///             c: WithdrawReasons,
+///             d: ExistenceRequirement,
+///         ) -> Result<Self::NegativeImbalance, DispatchError>;
+///         fn make_free_balance_be(
+///             a: &T::AccountId,
+///             b: Self::Balance,
+///         ) -> SignedImbalance<Self::Balance, Self::PositiveImbalance>
+///     }
+///     trait ReservableCurrency<AccountId> {
+///         fn can_reserve(a: &T::AccountId, b: Self::Balance) -> bool;
+///         fn slash_reserved(
+///             a: &T::AccountId,
+///             b: Self::Balance,
+///         ) -> (Self::NegativeImbalance, Self::Balance);
+///         fn reserved_balance(a: &T::AccountId) -> Self::Balance;
+///         fn reserve(a: &T::AccountId, b: Self::Balance) -> DispatchResult;
+///         fn unreserve(a: &T::AccountId, b: Self::Balance) -> Self::Balance;
+///         fn repatriate_reserved(
+///             a: &T::AccountId,
+///             b: &T::AccountId,
+///             c: Self::Balance,
+///             d: BalanceStatus,
+///         ) -> Result<Self::Balance, DispatchError>;
+///     }
+/// }
+/// ```
```

Comment:
I'm not sure if this is a selling point or could create more confusing given the number of methods that should be mocked. Just wanted to show a complex example.